### PR TITLE
Generate a Java token kind enum with a mapping to the integer kinds.

### PR DESCRIPTION
**Pull request by matozoid**
@ftomassetti please check if this is what you need, and if the naming is good.

### CODE REVIEW

File: javaparser-core-generators/src/main/java/com/github/javaparser/generator/core/CoreGenerator.java [hunk 1]
```diff
@@ -11,28 +12,37 @@
 
 /**
  * Generates all generated visitors in the javaparser-core module.
+ * Suggested usage is by running the run_core_generators.sh script.
+ * You may want to run_metamodel_generator.sh before that.
  */
 public class CoreGenerator {
+    private static final ParserConfiguration parserConfiguration = new ParserConfiguration()
+//                                .setStoreTokens(false)
```
Comment (ftomassetti) [inline javaparser-core-generators/src/main/java/com/github/javaparser/generator/core/CoreGenerator.java:20]: Why do we have stuff commented out in the file?

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 2]
```diff
@@ -51,41 +55,36 @@ public JavaToken(int kind, String text) {
...
-        String text = token.image;
-
         // You could be puzzled by the following lines
-        //
+        // 
         // The reason why these lines are necessary is the fact that Java is ambiguous. There are cases where the
         // sequence of characters ">>>" and ">>" should be recognized as the single tokens ">>>" and ">>". In other
         // cases however we want to split those characters in single GT tokens (">").
-        //
+        //
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:63]: Maybe we do not want these extra spaces. Not that it really matters :)

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 3]
```diff
@@ -248,6 +238,315 @@ public boolean isOperator() {
...
..., REQUIRES(64), TO(65), WITH(66), OPEN(67), OPENS(68), USES(69), MODULE(70), EXPORTS(71), PROVIDES(72), TRANSITIVE(73), LONG_LITERAL(74), INTEGER_LITERAL(75), DECIMAL_LITERAL(76), HEX_LITERAL(77), OCTAL_LITERAL(78), BINARY_LITERAL(79), FLOATING_POINT_LITERAL(80), DECIMAL_FLOATING_POINT_LITERAL(81), DECIMAL_EXPONENT(82), HEXADECIMAL_FLOATING_POINT_LITERAL(83), HEXADECIMAL_EXPONENT(84), HEX_DIGITS(85), UNICODE_ESCAPE(86), CHARACTER_LITERAL(87), STRING_LITERAL(88), IDENTIFIER(89), LETTER(90), PART_LETTER(91), LPAREN(92), RPAREN(93), LBRACE(94), RBRACE(95), LBRACKET(96), RBRACKET(97), SEMICOLON(98), COMMA(99), DOT(100), AT(101), ASSIGN(102), LT(103), BANG(104), TILDE(105), HOOK(106), COLON(107), EQ(108), LE(109), GE(110), NE(111), SC_OR(112), SC_AND(113), INCR(114), DECR(115), PLUS(116), MINUS(117), STAR(118), SLASH(119), BIT_AND(120), BIT_OR(121), XOR(122), REM(123), LSHIFT(124), PLUSASSIGN(125), MINUSASSIGN(126), STARASSIGN(127), SLASHASSIGN(128), ANDASSIGN(129), ORASSIGN(130), XORASSIGN(131), REMASSIGN(132), LSHIFTASSIGN(133), RSIGNEDSHIFTASSIGN(134), RUNSIGNEDSHIFTASSIGN(135), ELLIPSIS(136), ARROW(137), DOUBLECOLON(138), RUNSIGNEDSHIFT(139), RSIGNEDSHIFT(140), GT(141), CTRL_Z(142);
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:244]: Could we generate it with one per line?

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 4]
```diff
@@ -248,6 +238,315 @@ public boolean isOperator() {
...
+
+        private final int kind;
+
+        Kind(int kind) {
+            this.kind = kind;
+        }
+
+        public Kind valueOf(int kind) {
+            switch(kind) {
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java]: I do not think we need this switch we can just iterate on Kind.values and find the one matching

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 5]
```diff
@@ -335,14 +633,15 @@ public int hashCode() {
...
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
         JavaToken javaToken = (JavaToken) o;
-
-        if (kind != javaToken.kind) return false;
-        if (!text.equals(javaToken.text)) return false;
-
+        if (kind != javaToken.kind)
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:641]: I personally always use braces with the if statement

File: javaparser-core/src/main/java/com/github/javaparser/ast/NodeList.java [hunk 6]
```diff
@@ -53,11 +53,7 @@
     private List<AstObserver> observers = new ArrayList<>();
 
     public NodeList() {
-        this((Node) null);
-    }
-
-    public NodeList(Node parent) {
-        setParentNode(parent);
+        parentNode = null;
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/ast/NodeList.java:56]: How is this related?

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 7]
```diff
@@ -248,6 +238,315 @@ public boolean isOperator() {
...
+
+        private final int kind;
+
+        Kind(int kind) {
+            this.kind = kind;
+        }
+
+        public Kind valueOf(int kind) {
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java]: should this be static?

**Review by ftomassetti — APPROVED** [review]
It looks good but I would drop the switch in valueof and make the method static

File: javaparser-core-generators/src/main/java/com/github/javaparser/generator/core/CoreGenerator.java (hunk 1 above)
Comment (matozoid) [reply javaparser-core-generators/src/main/java/com/github/javaparser/generator/core/CoreGenerator.java:20]: See #1301

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java (hunk 2 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:63]: Everything gets reformatted because this file is now part of the code generation step. See also #1301 :-)

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java (hunk 3 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:244]: Everything gets reformatted because this file is now part of the code generation step. See also #1301 :-)

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java (hunk 7 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/JavaToken.java]: I think it should, good one!

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java (hunk 4 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/JavaToken.java]: I'm now thinking it could have been a single array access if we generate that array instead. Hmmm.

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java (hunk 5 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:641]: Nothing happened here but reformatting of IDE generated code.

File: javaparser-core/src/main/java/com/github/javaparser/ast/NodeList.java (hunk 6 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/ast/NodeList.java]: Nothing happened here but reformatting of IDE generated code.

Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/ast/NodeList.java:56]: I wanted to create a nodelist with a single item, but very confusingly, if you have only one item it becomes the parent instead of an item?!? This constructor wasn't used anywhere so I removed it. Deprecation would mean that it got in the way for longer. Hopefully nobody depends on this behaviour.
